### PR TITLE
Added $ Materials / Layers / Constructions block, and related translations

### PR DESCRIPTION
**Pull request by TfedUD**
May have not explicitly used @dataclass() 'Correctly'..
will circle back and refactor everything completely cleanly to 3.xx syntax, i.e f'string{self.formatting}' : but wanted to make forward progress; then circle back and clean

### CODE REVIEW

**Comment by mostaphaRoudsari** [comment]
Hi @TfedUD! This PR has two issues:

1. It is not reset to ladybug-tools/master and as a result, it includes all the commits that are already merged.
2. It has conflicting files.

You should start using a feature branch to make it easier for yourself. This is the process:

1. Reset your forks master to the lastest ladybug-tools/master
2. Create a new branch for your new feature
3. Work on your new feature
4. If during your development new commits are being merged into the ladybug-tools/master rebase your feature branch with the latest and resolve any possible conflicts.
5. Send a PR!
6. Address the issue until the PR is merged.
7. Delete the feature branch and go back to 1 for the next feature.

Someone from our team can probably show you how to do this. Meanwhile, I will rebase your current PR and push a fix.

File: dragonfly_doe2/doe/mats_layers_constr.py [hunk 1]
```diff
@@ -1,14 +1,16 @@
...
 
 
 class Material(object):
     """Material Obj
     refer to:
-        assets\DOE22Vol2-Dictionary_48r.pdf pg: 97
+        assets\\DOE22Vol2-Dictionary_48r.pdf pg: 97
     """
 
     def __init__(self, _hbmat):
```
Comment (mostaphaRoudsari) [inline dragonfly_doe2/doe/mats_layers_constr.py:16]: Unless you really need it, I don't think `hbmaterial` should be a property here. Similar to other classes you should add a `classmethod` for `from_hb_material`.

File: dragonfly_doe2/doe/mats_layers_constr.py [hunk 2]
```diff
@@ -42,43 +44,91 @@ def lbt_to_ip(self, _val, _to, _from):
...
+        # TODO: for sake of OCD: revert to 3.x f'strings'
         if self.mat_type == 'PROPERTIES':
-            return'"{display_name}" = MATERIAL\n'.format(
+
+            return'\n"{display_name}" = MATERIAL\n'.format(
                 display_name=self.hbmat.display_name) + \
                 '   TYPE            = {mat_type}\n'.format(mat_type=self.mat_type) + \
                 '   THICKNESS       = {thickness}\n'.format(
-                    thickness=self.lbt_to_ip(self.hbmat.thickness, 'ft', 'm')) + \
+                    thickness=self.lbt_to_ip([self.hbmat.thickness], 'ft', 'm')) + \
```
Comment (mostaphaRoudsari) [inline dragonfly_doe2/doe/mats_layers_constr.py:55]: As I mentioned I won't reference the hbmaterial here - you can build the class once and then reuse it. I will write a sample code and send a PR for you to review.

File: dragonfly_doe2/doe/mats_layers_constr.py [hunk 3]
```diff
@@ -42,43 +44,91 @@ def lbt_to_ip(self, _val, _to, _from):
...
-    via __repr__
-    TODO: make it happen! easy day! tomorrow.
+    via __repr__.
+
+        Returns:
+          $  Materials / Layers / Constructions *.inp block
     """
+    inp_constructions: list = None
+    inp_layers: list = None
+    inp_materials: list = None
```
Comment (mostaphaRoudsari) [inline dragonfly_doe2/doe/mats_layers_constr.py:85]: We already know this is an inp object. We should keep the names clean.

```suggestion
    constructions: list = None
    layers: list = None
    materials: list = None
```

Also aren't these inputs required? Why are you putting default None values for them?

File: dragonfly_doe2/doe/mats_layers_constr.py [hunk 4]
```diff
@@ -42,43 +44,91 @@ def lbt_to_ip(self, _val, _to, _from):
...
 
-    # ...STATICMETHOD!! noo... nooo... ...maybe. unitl tomoz!
+        cls_.inp_constructions = inp_cons
+        cls_.inp_layers = inp_layrs
+        cls_.inp_materials = inp_mats
+        return(cls_)
+
+    def to_inp(self):
+        # ? if these could be list comps that would be chill,
+        # TODO: Look at replacing with lambda func
```
Comment (mostaphaRoudsari) [inline dragonfly_doe2/doe/mats_layers_constr.py:122]: Why? 🙃 - What's wrong with normal Python functions?

File: dragonfly_doe2/doe/mats_layers_constr.py [hunk 5]
```diff
@@ -42,43 +44,91 @@ def lbt_to_ip(self, _val, _to, _from):
...
+
+        block = fb.matslayers
+        for mat in self.inp_materials:
+            block += mat.to_inp()
```
Comment (mostaphaRoudsari) [inline dragonfly_doe2/doe/mats_layers_constr.py:126]: Concatenating strings is not a good practice. Always create a list and join them.

File: dragonfly_doe2/doe/mats_layers_constr.py [hunk 6]
```diff
@@ -42,43 +44,91 @@ def lbt_to_ip(self, _val, _to, _from):
...
+        for lay in self.inp_layers:
+            block += lay
+        for cons in self.inp_constructions:
+            block += cons
+        return(block)
```
Comment (mostaphaRoudsari) [inline dragonfly_doe2/doe/mats_layers_constr.py:131]: I don't know why you do this but it making me crazy! 😄 - no `( )` is needed

```suggestion
        return block
```

File: tests/writer_test.py [hunk 7]
```diff
@@ -6,7 +6,7 @@
 
 def test_dfjson_translate():
     """Test translating a DFJSON file to an inp file."""
-    df_json = 'assets/reference_dfm/simple_example_v1.dfjson'
+    df_json = 'assets/reference_dfm/complex_model.dfjson'
```
Comment (mostaphaRoudsari) [inline tests/writer_test.py:9]: Brave! 😄

File: dragonfly_doe2/doe/model.py [hunk 8]
```diff
@@ -32,8 +34,11 @@ def from_df_model(cls, df_model: DFModel, run_period=None):
             polygons.append(Polygon.from_story(story))
             for room in story:
                 polygons.append(Polygon.from_room(room))
+        mats_layers_constrs = MatsLayersConstructions().from_hb_constructions(
+            df_model.properties.energy.constructions)
 
-        return cls(df_model.display_name, run_period, polygons=polygons)
+        return cls(df_model.display_name, run_period, polygons=polygons,
+                   mats_layers_constrs=mats_layers_constrs)
```
Comment (mostaphaRoudsari) [inline dragonfly_doe2/doe/model.py:41]: Nicely done!

**Review by mostaphaRoudsari — COMMENTED** [review]
Great start @TfedUD! I'll send a PR to your branch to share some sample codes based on my comments.

File: dragonfly_doe2/doe/mats_layers_constr.py (hunk 5 above)
Comment (TfedUD) [reply dragonfly_doe2/doe/mats_layers_constr.py:126]: ohhhh!! Right on, I typically do that; I thought this was a little 'cleaner' and less verbose for some reason; I'll look into the details on that!

File: dragonfly_doe2/doe/mats_layers_constr.py (hunk 3 above)
Comment (TfedUD) [reply dragonfly_doe2/doe/mats_layers_constr.py:85]: I wasn't quite sure, tbh:
in the 'simple.dfjson'; there are no constructions/mats etc; so the intent was to basically have a 'None' return for this block; so would still get the '$  mats/layers/constrs' inp block sans the actual inputs if desired by user.
(if curious: JSON editor extension nickdemayo.vscode-json-editor top 3 rando extensions imo; up there with the pdf viewer)
image

File: dragonfly_doe2/doe/mats_layers_constr.py (hunk 4 above)
Comment (TfedUD) [reply dragonfly_doe2/doe/mats_layers_constr.py:122]: 😄 I picked up the habbit from Ed giving me feedback and input on code to 'if can make multi-line functions one-liners via list comp etc: do it'.

File: dragonfly_doe2/doe/mats_layers_constr.py (hunk 5 above)
Comment (TfedUD) [reply dragonfly_doe2/doe/mats_layers_constr.py:126]: OHHH that's really good to know.. I'd been habitually doing yee-olde ``` '\n'.join(thing for thing in things) ```,  but picked this up during some rabbit hole on stackoverflow some time recently...; thought it to be less verbose.
tbh I didn't know this was a thing lol.
I'm used to only ever ```+=``` for:
```python
count = 0
while count < 5:
   do_stuff( )
   count += 1
# etc
```

File: dragonfly_doe2/doe/mats_layers_constr.py (hunk 6 above)
Comment (TfedUD) [reply dragonfly_doe2/doe/mats_layers_constr.py:131]: 10:4 solid copy!! I started doing this a while ago; and not even consistently, which makes it extra annoying imo 😅.
Will make extra endeavor to ```STOP``` !

File: dragonfly_doe2/doe/mats_layers_constr.py (hunk 1 above)
Comment (TfedUD) [reply dragonfly_doe2/doe/mats_layers_constr.py:16]: 10:4! will re-fac to operate off the @dataclass() MO, am liking that alot!

File: dragonfly_doe2/doe/mats_layers_constr.py (hunk 5 above)
Comment (mostaphaRoudsari) [reply dragonfly_doe2/doe/mats_layers_constr.py]: Integers are different from strings

**Pull request merged**